src/mysql_protocol.erl

```diff
@@ -1220,6 +1214,96 @@ recv_packet(SockModule, Socket, Timeout, ExpectSeqNum, Acc) ->
+        false ->
+            {{error, not_allowed}, SeqNum0};
+        {ok, Handle} ->
+            {ok, SeqNum2} = send_file_chunk(SockModule, Socket, Handle, SeqNum0),
+            ok = file:close(Handle),
+            {ok, SeqNum2};
+        {error, _Reason} = E ->
+            {E, SeqNum0}
+    end,
+    {ok, SeqNum3} = send_packet(SockModule, Socket, <<>>, SeqNum1),
+    {Result, SeqNum3}.
+
+-spec is_allowed_path(file:name_all(), [file:name_all()]) -> boolean().
+is_allowed_path(Path, AllowedPaths) ->
+    Path1 = normalize_split_path(Path),
+    lists:any(
+        fun (AllowedPath) ->
+            AllowedPath1 = normalize_split_path(AllowedPath),
+            is_prefix_path(AllowedPath1, Path1)
+        end,
+        AllowedPaths
+    ).
+
+is_prefix_path([], _Path2) ->
+    true;
+is_prefix_path(_Path1, []) ->
+    false;
+is_prefix_path([Segment1|Path1], Path2=[Segment2|_]) when is_binary(Segment1), is_list(Segment2) ->
+    is_prefix_path([binary_to_list(Segment1)|Path1], Path2);
+is_prefix_path(Path1=[Segment1|_], [Segment2|Path2]) when is_list(Segment1), is_binary(Segment2) ->
+    is_prefix_path(Path1, [binary_to_list(Segment2)|Path2]);
+is_prefix_path([Segment1|Path1], [Segment2|Path2]) ->
+    case string:equal(Segment1, Segment2) of
+        true -> is_prefix_path(Path1, Path2);
+        false -> false
+    end.
+
+-spec normalize_split_path(file:name_all()) -> list().
+normalize_split_path(Path) ->
+    Path1 = filename:flatten(Path),
+    Path2 = case filename:pathtype(Path1) of
+        absolute ->
+            Path1;
+        relative ->
+            {ok, Cwd} = file:get_cwd(),
+            filename:absname_join(Cwd, Path1);
+        volumerelative ->
+            Volume = hd(filename:split(Path1)),
+            {ok, Cwd} = file:get_cwd(Volume),
+            filename:absname_join(Cwd, Path1)
+    end,
+    [Root|RelPath] = filename:split(Path2),
+    normalize_split_path(Root, RelPath, []).
+
+normalize_split_path(Root, [], Acc) ->
+    [Root|lists:reverse(Acc)];
+normalize_split_path(Root, ["."|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [<<".">>|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [".."|More], []) ->
+    normalize_split_path(Root, More, []);
+normalize_split_path(Root, [<<"..">>|More], []) ->
+    normalize_split_path(Root, More, []);
+normalize_split_path(Root, [".."|More], [_|Acc]) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [<<"..">>|More], [_|Acc]) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [Segment|More], Acc) ->
+    normalize_split_path(Root, More, [Segment|Acc]).
+
+-spec send_file_chunk(module(), term(), Handle :: file:io_device(), SeqNum :: integer()) ->
+    {ok, NextSeqNum :: integer()}.
+send_file_chunk(SockModule, Socket, Handle, SeqNum0) ->
+    case file:read(Handle, 16#ffffff) of
+        eof ->
+            {ok, SeqNum0};
+        {ok, <<>>} ->
```

Comment:
That's a precautionary thing. While I have never seen it happen, the documentation does not say that it _can't _ return empty. And if it does return an empty binary but it is not on eof, it will send an empty packet to the server, indicating eof to the server while the file has actually not been fully sent.